### PR TITLE
feat: Remember last used output paths and take selection for job submissions

### PR DESCRIPTION
**Pull request by gianelli-lagos-aws**
Fixes: *< (https://github.com/aws-deadline/deadline-cloud-for-cinema-4d/issues/95) >*

### What was the problem/requirement? (What/Why)
The fields output_path, multi_pass_path, and take_selection in the submitter's job specific settings do not have sticky settings which can be inconvenient. The fields override_output_path and override_multi_pass_path do not exist and they would be useful in order to save checkbox checked states. 

### What was the solution? (How)
Add sticky settings to output_path, multi_pass_output, and take_selection. 
Create two new boolean fields, override_output_path and override_multi_pass_path, with sticky metadata to track checkbox states. For override_output_path and override_multi_pass_path, if the checkbox is unchecked the default render settings paths will be used for the job. 

### What is the impact of this change?
Customers will have settings that are saved and reused between submissions.

### How was this change tested?
* Built the package with the changes and verified that settings persist after closing and reopening Cinema 4D.
* Tested with existing settings files to ensure backward compatibility is maintained.
* Examined the JSON settings file to confirm that the new fields are being properly saved.
* Tested edge cases such as enabling/disabling checkboxes and changing paths multiple times.
* Ran unit tests:
  * 0.02s call     test/unit/deadline_adaptor_for_cinema4d/Cinema4DAdaptor/test_adaptor.py::TestCinema4DAdaptor_on_cleanup::test_handle_errors_on_error_stdout[Project not found-True]
============================================================ 53 passed in 4.58s ============================================================ 
* Ran integration tests:
  * =========================================================== slowest 5 durations ============================================================ 
109.25s call     test/integ/test_cinema4d.py::test_integ[redshift_textured_with_nonascii_characters]
82.20s call     test/integ/test_cinema4d.py::test_integ[redshift_takes]
78.92s call     test/integ/test_cinema4d.py::test_integ[redshift_textured]
75.83s call     test/integ/test_cinema4d.py::test_integ[redshift]
54.39s call     test/integ/test_cinema4d.py::test_integ[physical]
====================================================== 6 passed in 456.20s (0:07:36) =======================================================

*By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.*

### CODE REVIEW

File: src/deadline/cinema4d_submitter/data_classes.py [hunk 1]
```diff
@@ -87,6 +89,8 @@ def load_sticky_settings(self, scene_filename: str):
                         if name in sticky_fields:
                             if name == "timeouts":
                                 self.timeouts.update_from_sticky_settings(value)
+                            elif name == "take_selection":
```
Comment (karthikbekalp) [inline src/deadline/cinema4d_submitter/data_classes.py:93]: Nit: A comment explaining why "take_selection" requires a separate check would be helpful.

File: src/deadline/cinema4d_submitter/cinema4d_render_submitter.py [hunk 2]
```diff
@@ -415,12 +415,20 @@ def create_job_bundle(
         submit_takes = takes["current_data_list"]
 
     # Add overrides to asset references and update the paths with C4D render path tokens.
-    if settings.output_path:
+    default_output_path, default_multi_pass_path = Scene.get_output_paths()
```
Comment (karthikbekalp) [inline src/deadline/cinema4d_submitter/cinema4d_render_submitter.py]: Nit: Instead of `default_output_path` can we name it as `scene_output_path` to emphasize the source of the path?

**Review by karthikbekalp — COMMENTED** [review]
Thanks Gianelli. The PR looks great albeit some nit comments. 

Can you also check with different render tokens and if those work as expected too? https://help.maxon.net/c4d/en-us/#html/54506.html

**Comment by karthikbekalp** [comment]
I think the commit message could be improved.

Instead of:
`feat: add sticky settings to job specific fields for cinema4d submitter`

We could use:
`feat: Remember last used output paths and take selection for job submissions`

This would be more helpful for our customers when they read the release notes to decide whether to upgrade to a newer release.

**Review by karthikbekalp — DISMISSED** [review]
LGTM. Great job!

File: src/deadline/cinema4d_submitter/cinema4d_render_submitter.py (hunk 2 above)
Comment (gianelli-lagos-aws) [reply src/deadline/cinema4d_submitter/cinema4d_render_submitter.py]: Yes, I think that `scene_output_path` is better. Thank you!

File: src/deadline/cinema4d_submitter/data_classes.py (hunk 1 above)
Comment (gianelli-lagos-aws) [reply src/deadline/cinema4d_submitter/data_classes.py:93]: I agree. Thank you!

Comment (gianelli-lagos-aws) [reply src/deadline/cinema4d_submitter/data_classes.py:93]: I agree. Thank you!

File: src/deadline/cinema4d_submitter/cinema4d_render_submitter.py (hunk 2 above)
Comment (gianelli-lagos-aws) [reply src/deadline/cinema4d_submitter/cinema4d_render_submitter.py]: Yes, i think `scene_output_path` is better. Thank you!

File: src/deadline/cinema4d_submitter/cinema4d_render_submitter.py [hunk 3]
```diff
@@ -415,12 +415,24 @@ def create_job_bundle(
...
 
     # Add overrides to asset references and update the paths with C4D render path tokens.
-    if settings.output_path:
-        settings.output_path = Scene.replace_render_path_tokens(settings.output_path)
-        asset_references.output_directories.add(os.path.dirname(settings.output_path))
-    if settings.multi_pass_path:
-        settings.multi_pass_path = Scene.replace_render_path_tokens(settings.multi_pass_path)
-        asset_references.output_directories.add(os.path.dirname(settings.multi_pass_path))
+    scene_output_path, scene_multi_pass_path = Scene.get_output_paths()
+    if settings.override_output_path:
```
Comment (joel-wong-aws) [inline src/deadline/cinema4d_submitter/cinema4d_render_submitter.py:419]: I think there's one case that this approach misses.

Imagine I set my output path via Cinema 4D (Render > Render Settings > Save) to C:\A\Path. I then open the submitter, check that my job settings show an output path of `C:\A\Path` (no path override) and submit my job.

The second time I submit, I change my output path to `C:\B\Path` in Cinema 4D (Render > Render Settings > Save). When I open the submitter the second time, my job settings will still show an output path of `C:\A\Path` with no path override, even though it will actually submit with an output path of `C:\B\Path`.

I think we need to handle this case to avoid customer confusion. There are two ways we can do this:
(1) If override output path is not set, then do not load the output path from the sticky settings file
(2) If override output path is not set, then do not save the output path to the sticky settings file

I think either approach could be okay. (1) saves the correct output path to the sticky settings file but doesn't use it, while (2) leaves an empty/missing output path

Comment (gianelli-lagos-aws) [reply src/deadline/cinema4d_submitter/cinema4d_render_submitter.py:419]: Hello, I can see how this approach misses! I think I will go for the second option since if override output path is false then it best for the output path to not be in the json file. Thank you!

File: src/deadline/cinema4d_submitter/data_classes.py [hunk 4]
```diff
@@ -113,7 +113,12 @@ def save_sticky_settings(self, scene_filename: str):
                 obj = {
                     field.name: getattr(self, field.name)
                     for field in dataclasses.fields(self)
-                    if field.metadata.get("sticky") and field.name != "timeouts"
+                    if field.metadata.get("sticky")
+                    and field.name != "timeouts"
+                    and not (
+                        (field.name == "output_path" and not self.override_output_path)
+                        or (field.name == "multi_pass_path" and not self.override_multi_pass_path)
+                    )
```
Comment (karthikbekalp) [inline src/deadline/cinema4d_submitter/data_classes.py]: nit: I think we can improve the readability of the code here:

```
             with open(sticky_settings_filename, "w", encoding="utf8") as fh:
                obj = {
                    field.name: getattr(self, field.name)
                    for field in dataclasses.fields(self)
                    if field.metadata.get("sticky")
                    and field.name not in ["timeouts", "output_path", "multi_pass_path"]
                }
                obj["timeouts"] = self.timeouts.to_sticky_settings_dict()
                if self.override_output_path:
                    obj["output_path"] = self.output_path
                if self.override_multi_pass_path:
                    obj["multi_pass_path"] = self.multi_pass_path
                json.dump(obj, fh, indent=1)
```

**Review by karthikbekalp — COMMENTED** [review]
Added one comment to improve readability, but other than that. LGTM.

File: src/deadline/cinema4d_submitter/data_classes.py (hunk 4 above)
Comment (gianelli-lagos-aws) [reply src/deadline/cinema4d_submitter/data_classes.py]: I agree, this is much clearer! I will work on that, thank you.

**Review by karthikbekalp — APPROVED** [review]

**Review by joel-wong-aws — APPROVED** [review]
